config/validation_test.go

```diff
@@ -0,0 +1,298 @@
+			},
+			"emptybackends": {
+				VolumeName:     "volume",
+				PolicyName:     "policy",
+				RuntimeOptions: RuntimeOptions{UseSnapshots: false},
+			},
+		},
+	}
+	PolicyConfigs = map[string]map[string]*Policy{
+		"valid": {
+			"basic": testPolicies["basic"],
+			"basicceph": {
+				Name: "basicceph",
+				Backends: &BackendDrivers{
+					CRUD:     "ceph",
+					Mount:    "ceph",
+					Snapshot: "ceph",
+				},
+			},
+			"basicnfs": {
+				Name: "basicnfs",
+				Backends: &BackendDrivers{
+					CRUD:     "",
+					Mount:    "nfs",
+					Snapshot: "",
+				},
+			},
+			"backendceph": {
+				Name:    "backendceph",
+				Backend: "ceph",
+			},
+			"backendnfs": {
+				Name:    "backendnfs",
+				Backend: "nfs",
+			},
+			"withbackendattr1": {
+				Name: "withbackendattr1",
+				Backends: &BackendDrivers{
+					CRUD:     "ceph",
+					Mount:    "ceph",
+					Snapshot: "ceph",
+				},
+				Backend: "nfs",
+			},
+			"withbackendattr2": {
+				Name: "withbackendattr2",
+				Backends: &BackendDrivers{
+					CRUD:     "",
+					Mount:    "nfs",
+					Snapshot: "",
+				},
+				Backend: "ceph",
+			},
+		},
+		"invalid": {
+			"emptyname": {
+				Name:    "",
+				Backend: "ceph",
+			},
+			"badbackend": {
+				Name:    "badbackend",
+				Backend: "bad",
+			},
+			"emptybackendconfig": {
+				Name: "emptybackendconfig",
+			},
+			"emptybackend": {
+				Name:    "emptybackend",
+				Backend: "", // omitempty
+			},
+			"emptybackendsconfig": {
+				Name: "emptybackendsconfig",
+				Backends: &BackendDrivers{ // this field will be omitted - "omitempty"
+					CRUD:     "",
+					Mount:    "",
+					Snapshot: "",
+				},
+			},
+			"badbackendsconfig": {
+				Name: "badbackendsconfig",
+				Backends: &BackendDrivers{
+					CRUD:     "bad",
+					Mount:    "bad",
+					Snapshot: "bad",
+				},
+			},
+			"nomount": {
+				Name:     "nomount",
+				Backends: &BackendDrivers{},
+			},
+			"nfsbackends": { //only "nfs" mount it is valid
+				Name: "crudnfs",
+				Backends: &BackendDrivers{
+					CRUD:     "nfs",
+					Mount:    "nfs",
+					Snapshot: "nfs",
+				},
+			},
+		},
+	}
+	RuntimeConfigs = map[string]map[string]*RuntimeOptions{
+		"valid": {
+			"snapshots": {
+				UseSnapshots: true,
+				Snapshot: SnapshotConfig{
+					Keep:      10,
+					Frequency: "1m",
+				},
+			},
+			"nosnapshots": {
+				UseSnapshots: false,
+			},
+		},
+		"invalid": {
+			"nosnapshotconfig": {
+				UseSnapshots: true, // requires snapshot configuration
+			},
+			"zerokeepvalue": {
+				UseSnapshots: true,
+				Snapshot: SnapshotConfig{
+					Keep:      0, // anything <1 is invalid
+					Frequency: "1m",
+				},
+			},
+			"emptyfrequency": {
+				UseSnapshots: true,
+				Snapshot: SnapshotConfig{
+					Keep:      1,
+					Frequency: "", // "" not allowed
+				},
+			},
+			"invalidfrequency": {
+				UseSnapshots: true,
+				Snapshot: SnapshotConfig{
+					Frequency: "badvalue", // accpets only (^[0-9]+.$)
```

Comment:
regexp engines are different between different systems. oniguruma is ruby's engine, which is very similar to perl's. Both are non-deterministic.

RE2, however.... which is the golang implementation, is different. It's a deterministic regexp engine. In this case it's the same, but in several cases you'd be incorrect (like egrep, or sed, or emacs).

Anyways my point. Specifically from https://golang.org/pkg/regexp/syntax/:

```
^              at beginning of text or line (flag m=true)
$              at end of text (like \z not \Z) or line (flag m=true)
\A             at beginning of text
\b             at ASCII word boundary (\w on one side and \W, \A, or \z on the other)
\B             not at ASCII word boundary
\z             at end of text
```

`\A` and `\z` only differ from `^` and `$` when `m` is applied. I think you'll find that this behavior is exactly the same in perl and ruby, but ruby turns `m` on by default.

The goal of this lecturing is to hopefully educate a little bit about how every single regexp implementation is different, and also express that this code should have no issues unless something in between sets `m`.